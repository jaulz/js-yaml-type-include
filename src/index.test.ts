import fs from 'fs'
import yaml from 'js-yaml'
import { createIncludeType, constructors } from './index'

const textFixture = fs.readFileSync('fixtures/include.txt', 'utf8')

describe('createType', () => {
  test('includes content of existing file', () => {
    const type = createIncludeType()
    const schema = new yaml.Schema({
      include: [yaml.DEFAULT_SAFE_SCHEMA],
      explicit: [type],
    })
    const content = yaml.load(
      `
includedFile: !!include "fixtures/include.txt"
  `,
      { schema }
    )

    expect(content).not.toBeNull()
    expect(content.includedFile).not.toBeNull()
    expect(content.includedFile).toEqual(textFixture)
    expect(content).toMatchSnapshot('load')
    expect(yaml.dump(content)).toMatchSnapshot('dump')
  })
  test('includes content of existing file with custom extensions', () => {
    const type = createIncludeType({
      extensions: [
        {
          pattern: /\.png$/,
          construct: constructors.createDataURI,
        },
      ],
      createNestedSchema: nestedType =>
        new yaml.Schema({
          include: [yaml.DEFAULT_SAFE_SCHEMA],
          explicit: [nestedType],
        }),
    })
    const schema = new yaml.Schema({
      include: [yaml.DEFAULT_SAFE_SCHEMA],
      explicit: [type],
    })
    const content = yaml.load(
      `
includedYAML: !!include "fixtures/include.yml"
includedPNG: !!include "fixtures/include.png"
  `,
      { schema }
    )

    expect(content).not.toBeNull()
    expect(content.includedYAML).not.toBeNull()
    expect(content.includedYAML.deepInclude.deep).toEqual(true)
    expect(content.includedPNG).not.toBeNull()
    expect(content).toMatchSnapshot('load')
    expect(yaml.dump(content)).toMatchSnapshot('dump')
  })

  test('includes content of existing file with custom path', () => {
    const type = createIncludeType({ relativeTo: 'fixtures/' })
    const schema = new yaml.Schema({
      include: [yaml.DEFAULT_SAFE_SCHEMA],
      explicit: [type],
    })
    const content = yaml.load(
      `
includedFile: !!include "include.txt"
  `,
      { schema }
    )

    expect(content).not.toBeNull()
    expect(content.includedFile).not.toBeNull()
    expect(content.includedFile).toEqual(textFixture)
    expect(content).toMatchSnapshot('load')
    expect(yaml.dump(content)).toMatchSnapshot('dump')
  })

  test('throw error for non-existing file', () => {
    const type = createIncludeType()
    const schema = new yaml.Schema({
      include: [yaml.DEFAULT_SAFE_SCHEMA],
      explicit: [type],
    })

    expect(() => {
      yaml.load(
        `
includedFile: !!include "fixtures/nonExistant.txt"
  `,
        { schema }
      )
    }).toThrowErrorMatchingSnapshot()
  })
})
