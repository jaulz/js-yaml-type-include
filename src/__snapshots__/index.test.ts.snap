// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createType includes content of existing file with custom extensions: dump 1`] = `
"includedYAML:
  includedObject:
    - true
    - 'no'
    - custom: content
  deepInclude:
    deep: true
    deepPNG: >-
      data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAUISURBVFhHtVdpTJRXFEWWQELSIXW01EqpjdUibnQ6oqioaEAWadQAkgiYFpywjEJpgWLskjaoSaX1ByGElEJTFlFBC5RAWfwhaCFAC0kTsK02hCUwwshuoJ7ee4chpimxzAw3OZn3vffuPWfufe/73rMywXYSLhN+JUzNo5OQSdhFWDbbRKi3tbVFaGgosrOzcePGDQG3w8LCYGdnB5rTQHBnB0vaAYI+KCgILS0tqKj8EdEaLXbtPSSI1iRQXxVaW1sRHBzMIh4TDrKjJWwjYVyr1aK7pwe7vQ8yAdRuCpw+5gYNQe3mJH1eew+gu7sbZ8+e5ecJghvBbGsKCAhAD5ErXlwN1RsKdJa/B/ScB/78xABqd5VHQ73BCS84rZK5nC3ybTaEMN0OOzg4oK2tDR6qnfBYr8BcZxrw4FOg6yOg9QMDuE19f9OvikRs89iB9vZ2sC/FCDCEMs0KeXFdu14mKX7wQwzw+8f4q0KDHvrH+CVVcJ/aD6mPxx5WnJa5pdeuIzw8nNvFHMgUW0HozsnJQUh4FLy3rgT+oHQTyakjm7Ff5WIoA8FH/SoiA91ljOfs36bE8dAI5ObmsoD7BGsOuFRTEHS8zdy3eyItYruh3u0pGK5LwEBNHNCRIuA29/EYz0mP8sCmrWrZohRjhODEAZdqKwn6W7duwW2LGudOvWUQYKz5b+eAtg8N4LZxTdCc8++qsHGzCuxLMXhLruKASzUWMFpTUwP/I8fhp37JUIK2FEzeScRo4xkDKUF/+wym7iTJGAvw93SGX+BR1NbWsgA9QckBl2oigIN8k5fPgTBUFw/0f4FL2n2wtbXGurVOeJ1gR+0LCd4ypqtPkLl53xagqanJfAFcx4mJCax+2QW+O9bIdhu7+z6+/zwQaac8kRrlie8+C4C+iTJAY4dpzhqXdZienkZERIT5AkpKSsD2U10dBbNBsJcLhqkEeHQRGJkHtXVUgnd2u2CFjT1+ptd1TEyMZIJgnoCrV69KBvgf8XpwfuU12FLgowdckaH1wgXCMR9X6Vvruh53792DRqMR8sTERFhbW1tGQF9fHyYnJ9Hb24uLl76Eeuc+KJ1dBW97euNy5tfQ6/WIjo4W8jrKWENDg+Uy0N/fLyKGhoYkG9w3ODgoePLkiZTp5MmTQl5WVibP7EvPlhPwLAYGBoR8ZGQEU1NTch6g+aiursbs7Kz4lJaWLp8Azsbw8LCQnzhxQsj5sFJeXr78Ap4lDwkJEfKKigrExsZKu6ioSEqwLAL+i9xYc14bkZGR0ldZWYnGxkbLCliMnFNuXKAzMzMLIuLi4szehnpOIwt4HrlRoFFEVFSUzCGY/DHiz/GjwsJC2f/PI3+2RP8SYfLnWA4kGRkZePr0qZDzcZz6FiVn8PZkzM3NISsri+ebfCBhK+IDKQv4v+Sjo6MLLyf2JZ8SiWSi+SsUCvj6+go5fxkXSzuXiC01NRV+fn7o6OiAvb09+5l1KGXjozVu3rwpBEZy/uW34NjYmIAFJCUlidDMzEzjBeUeBzDX3qStNBEfH4/x8fGFLanT6VBfXy/HruLiYiiVSnC28vPzkZyczOSTBL7OWcT4mvXYx8cHzc3N8sJhS0tLw549e1BVVYX09HR5Dc+vlTHCIXa0pPGFs9HGxgb+/v7g3VFQUIC8vDxcuXJFTj+Ojo5MfpuwhR2Wy7wIXxH4Ss5X82lC13zfbsISzMrqH5aPFsMMryCrAAAAAElFTkSuQmCC
includedPNG: >-
  data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAUISURBVFhHtVdpTJRXFEWWQELSIXW01EqpjdUibnQ6oqioaEAWadQAkgiYFpywjEJpgWLskjaoSaX1ByGElEJTFlFBC5RAWfwhaCFAC0kTsK02hCUwwshuoJ7ee4chpimxzAw3OZn3vffuPWfufe/73rMywXYSLhN+JUzNo5OQSdhFWDbbRKi3tbVFaGgosrOzcePGDQG3w8LCYGdnB5rTQHBnB0vaAYI+KCgILS0tqKj8EdEaLXbtPSSI1iRQXxVaW1sRHBzMIh4TDrKjJWwjYVyr1aK7pwe7vQ8yAdRuCpw+5gYNQe3mJH1eew+gu7sbZ8+e5ecJghvBbGsKCAhAD5ErXlwN1RsKdJa/B/ScB/78xABqd5VHQ73BCS84rZK5nC3ybTaEMN0OOzg4oK2tDR6qnfBYr8BcZxrw4FOg6yOg9QMDuE19f9OvikRs89iB9vZ2sC/FCDCEMs0KeXFdu14mKX7wQwzw+8f4q0KDHvrH+CVVcJ/aD6mPxx5WnJa5pdeuIzw8nNvFHMgUW0HozsnJQUh4FLy3rgT+oHQTyakjm7Ff5WIoA8FH/SoiA91ljOfs36bE8dAI5ObmsoD7BGsOuFRTEHS8zdy3eyItYruh3u0pGK5LwEBNHNCRIuA29/EYz0mP8sCmrWrZohRjhODEAZdqKwn6W7duwW2LGudOvWUQYKz5b+eAtg8N4LZxTdCc8++qsHGzCuxLMXhLruKASzUWMFpTUwP/I8fhp37JUIK2FEzeScRo4xkDKUF/+wym7iTJGAvw93SGX+BR1NbWsgA9QckBl2oigIN8k5fPgTBUFw/0f4FL2n2wtbXGurVOeJ1gR+0LCd4ypqtPkLl53xagqanJfAFcx4mJCax+2QW+O9bIdhu7+z6+/zwQaac8kRrlie8+C4C+iTJAY4dpzhqXdZienkZERIT5AkpKSsD2U10dBbNBsJcLhqkEeHQRGJkHtXVUgnd2u2CFjT1+ptd1TEyMZIJgnoCrV69KBvgf8XpwfuU12FLgowdckaH1wgXCMR9X6Vvruh53792DRqMR8sTERFhbW1tGQF9fHyYnJ9Hb24uLl76Eeuc+KJ1dBW97euNy5tfQ6/WIjo4W8jrKWENDg+Uy0N/fLyKGhoYkG9w3ODgoePLkiZTp5MmTQl5WVibP7EvPlhPwLAYGBoR8ZGQEU1NTch6g+aiursbs7Kz4lJaWLp8Azsbw8LCQnzhxQsj5sFJeXr78Ap4lDwkJEfKKigrExsZKu6ioSEqwLAL+i9xYc14bkZGR0ldZWYnGxkbLCliMnFNuXKAzMzMLIuLi4szehnpOIwt4HrlRoFFEVFSUzCGY/DHiz/GjwsJC2f/PI3+2RP8SYfLnWA4kGRkZePr0qZDzcZz6FiVn8PZkzM3NISsri+ebfCBhK+IDKQv4v+Sjo6MLLyf2JZ8SiWSi+SsUCvj6+go5fxkXSzuXiC01NRV+fn7o6OiAvb09+5l1KGXjozVu3rwpBEZy/uW34NjYmIAFJCUlidDMzEzjBeUeBzDX3qStNBEfH4/x8fGFLanT6VBfXy/HruLiYiiVSnC28vPzkZyczOSTBL7OWcT4mvXYx8cHzc3N8sJhS0tLw549e1BVVYX09HR5Dc+vlTHCIXa0pPGFs9HGxgb+/v7g3VFQUIC8vDxcuXJFTj+Ojo5MfpuwhR2Wy7wIXxH4Ss5X82lC13zfbsISzMrqH5aPFsMMryCrAAAAAElFTkSuQmCC
"
`;

exports[`createType includes content of existing file with custom extensions: load 1`] = `
Object {
  "includedPNG": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAUISURBVFhHtVdpTJRXFEWWQELSIXW01EqpjdUibnQ6oqioaEAWadQAkgiYFpywjEJpgWLskjaoSaX1ByGElEJTFlFBC5RAWfwhaCFAC0kTsK02hCUwwshuoJ7ee4chpimxzAw3OZn3vffuPWfufe/73rMywXYSLhN+JUzNo5OQSdhFWDbbRKi3tbVFaGgosrOzcePGDQG3w8LCYGdnB5rTQHBnB0vaAYI+KCgILS0tqKj8EdEaLXbtPSSI1iRQXxVaW1sRHBzMIh4TDrKjJWwjYVyr1aK7pwe7vQ8yAdRuCpw+5gYNQe3mJH1eew+gu7sbZ8+e5ecJghvBbGsKCAhAD5ErXlwN1RsKdJa/B/ScB/78xABqd5VHQ73BCS84rZK5nC3ybTaEMN0OOzg4oK2tDR6qnfBYr8BcZxrw4FOg6yOg9QMDuE19f9OvikRs89iB9vZ2sC/FCDCEMs0KeXFdu14mKX7wQwzw+8f4q0KDHvrH+CVVcJ/aD6mPxx5WnJa5pdeuIzw8nNvFHMgUW0HozsnJQUh4FLy3rgT+oHQTyakjm7Ff5WIoA8FH/SoiA91ljOfs36bE8dAI5ObmsoD7BGsOuFRTEHS8zdy3eyItYruh3u0pGK5LwEBNHNCRIuA29/EYz0mP8sCmrWrZohRjhODEAZdqKwn6W7duwW2LGudOvWUQYKz5b+eAtg8N4LZxTdCc8++qsHGzCuxLMXhLruKASzUWMFpTUwP/I8fhp37JUIK2FEzeScRo4xkDKUF/+wym7iTJGAvw93SGX+BR1NbWsgA9QckBl2oigIN8k5fPgTBUFw/0f4FL2n2wtbXGurVOeJ1gR+0LCd4ypqtPkLl53xagqanJfAFcx4mJCax+2QW+O9bIdhu7+z6+/zwQaac8kRrlie8+C4C+iTJAY4dpzhqXdZienkZERIT5AkpKSsD2U10dBbNBsJcLhqkEeHQRGJkHtXVUgnd2u2CFjT1+ptd1TEyMZIJgnoCrV69KBvgf8XpwfuU12FLgowdckaH1wgXCMR9X6Vvruh53792DRqMR8sTERFhbW1tGQF9fHyYnJ9Hb24uLl76Eeuc+KJ1dBW97euNy5tfQ6/WIjo4W8jrKWENDg+Uy0N/fLyKGhoYkG9w3ODgoePLkiZTp5MmTQl5WVibP7EvPlhPwLAYGBoR8ZGQEU1NTch6g+aiursbs7Kz4lJaWLp8Azsbw8LCQnzhxQsj5sFJeXr78Ap4lDwkJEfKKigrExsZKu6ioSEqwLAL+i9xYc14bkZGR0ldZWYnGxkbLCliMnFNuXKAzMzMLIuLi4szehnpOIwt4HrlRoFFEVFSUzCGY/DHiz/GjwsJC2f/PI3+2RP8SYfLnWA4kGRkZePr0qZDzcZz6FiVn8PZkzM3NISsri+ebfCBhK+IDKQv4v+Sjo6MLLyf2JZ8SiWSi+SsUCvj6+go5fxkXSzuXiC01NRV+fn7o6OiAvb09+5l1KGXjozVu3rwpBEZy/uW34NjYmIAFJCUlidDMzEzjBeUeBzDX3qStNBEfH4/x8fGFLanT6VBfXy/HruLiYiiVSnC28vPzkZyczOSTBL7OWcT4mvXYx8cHzc3N8sJhS0tLw549e1BVVYX09HR5Dc+vlTHCIXa0pPGFs9HGxgb+/v7g3VFQUIC8vDxcuXJFTj+Ojo5MfpuwhR2Wy7wIXxH4Ss5X82lC13zfbsISzMrqH5aPFsMMryCrAAAAAElFTkSuQmCC",
  "includedYAML": Object {
    "deepInclude": Object {
      "deep": true,
      "deepPNG": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAUISURBVFhHtVdpTJRXFEWWQELSIXW01EqpjdUibnQ6oqioaEAWadQAkgiYFpywjEJpgWLskjaoSaX1ByGElEJTFlFBC5RAWfwhaCFAC0kTsK02hCUwwshuoJ7ee4chpimxzAw3OZn3vffuPWfufe/73rMywXYSLhN+JUzNo5OQSdhFWDbbRKi3tbVFaGgosrOzcePGDQG3w8LCYGdnB5rTQHBnB0vaAYI+KCgILS0tqKj8EdEaLXbtPSSI1iRQXxVaW1sRHBzMIh4TDrKjJWwjYVyr1aK7pwe7vQ8yAdRuCpw+5gYNQe3mJH1eew+gu7sbZ8+e5ecJghvBbGsKCAhAD5ErXlwN1RsKdJa/B/ScB/78xABqd5VHQ73BCS84rZK5nC3ybTaEMN0OOzg4oK2tDR6qnfBYr8BcZxrw4FOg6yOg9QMDuE19f9OvikRs89iB9vZ2sC/FCDCEMs0KeXFdu14mKX7wQwzw+8f4q0KDHvrH+CVVcJ/aD6mPxx5WnJa5pdeuIzw8nNvFHMgUW0HozsnJQUh4FLy3rgT+oHQTyakjm7Ff5WIoA8FH/SoiA91ljOfs36bE8dAI5ObmsoD7BGsOuFRTEHS8zdy3eyItYruh3u0pGK5LwEBNHNCRIuA29/EYz0mP8sCmrWrZohRjhODEAZdqKwn6W7duwW2LGudOvWUQYKz5b+eAtg8N4LZxTdCc8++qsHGzCuxLMXhLruKASzUWMFpTUwP/I8fhp37JUIK2FEzeScRo4xkDKUF/+wym7iTJGAvw93SGX+BR1NbWsgA9QckBl2oigIN8k5fPgTBUFw/0f4FL2n2wtbXGurVOeJ1gR+0LCd4ypqtPkLl53xagqanJfAFcx4mJCax+2QW+O9bIdhu7+z6+/zwQaac8kRrlie8+C4C+iTJAY4dpzhqXdZienkZERIT5AkpKSsD2U10dBbNBsJcLhqkEeHQRGJkHtXVUgnd2u2CFjT1+ptd1TEyMZIJgnoCrV69KBvgf8XpwfuU12FLgowdckaH1wgXCMR9X6Vvruh53792DRqMR8sTERFhbW1tGQF9fHyYnJ9Hb24uLl76Eeuc+KJ1dBW97euNy5tfQ6/WIjo4W8jrKWENDg+Uy0N/fLyKGhoYkG9w3ODgoePLkiZTp5MmTQl5WVibP7EvPlhPwLAYGBoR8ZGQEU1NTch6g+aiursbs7Kz4lJaWLp8Azsbw8LCQnzhxQsj5sFJeXr78Ap4lDwkJEfKKigrExsZKu6ioSEqwLAL+i9xYc14bkZGR0ldZWYnGxkbLCliMnFNuXKAzMzMLIuLi4szehnpOIwt4HrlRoFFEVFSUzCGY/DHiz/GjwsJC2f/PI3+2RP8SYfLnWA4kGRkZePr0qZDzcZz6FiVn8PZkzM3NISsri+ebfCBhK+IDKQv4v+Sjo6MLLyf2JZ8SiWSi+SsUCvj6+go5fxkXSzuXiC01NRV+fn7o6OiAvb09+5l1KGXjozVu3rwpBEZy/uW34NjYmIAFJCUlidDMzEzjBeUeBzDX3qStNBEfH4/x8fGFLanT6VBfXy/HruLiYiiVSnC28vPzkZyczOSTBL7OWcT4mvXYx8cHzc3N8sJhS0tLw549e1BVVYX09HR5Dc+vlTHCIXa0pPGFs9HGxgb+/v7g3VFQUIC8vDxcuXJFTj+Ojo5MfpuwhR2Wy7wIXxH4Ss5X82lC13zfbsISzMrqH5aPFsMMryCrAAAAAElFTkSuQmCC",
    },
    "includedObject": Array [
      true,
      "no",
      Object {
        "custom": "content",
      },
    ],
  },
}
`;

exports[`createType includes content of existing file with custom path: dump 1`] = `
"includedFile: Custom Include
"
`;

exports[`createType includes content of existing file with custom path: load 1`] = `
Object {
  "includedFile": "Custom Include",
}
`;

exports[`createType includes content of existing file: dump 1`] = `
"includedFile: Custom Include
"
`;

exports[`createType includes content of existing file: load 1`] = `
Object {
  "includedFile": "Custom Include",
}
`;

exports[`createType throw error for non-existing file 1`] = `
"cannot resolve a node with !<tag:yaml.org,2002:include> explicit tag at line 2, column 51:
     ... clude \\"fixtures/nonExistant.txt\\"
                                         ^"
`;
